src/db/gc/mod.rs

```diff
@@ -0,0 +1,166 @@
+            gc_receiver,
+        }
+    }
+    // Populate the initial set with all the available database keys.
+    fn populate(&mut self) -> anyhow::Result<()> {
+        self.marked = self.db.get_keys()?;
+        Ok(())
+    }
+
+    // Filter out the initial set, leaving only the entries that need to be removed.
+    // NOTE: One concern here is that this is going to consume a lot of CPU.
+    async fn filter(&mut self, tipset: Arc<Tipset>, depth: ChainEpochDelta) -> anyhow::Result<()> {
+        // NOTE: We want to keep all the block headers from genesis to heaviest tipset epoch.
+        let mut stream = stream_graph(
+            self.db.clone(),
+            (*tipset).clone().chain(self.db.clone()),
+            depth,
+        );
+
+        while let Some(block) = stream.next().await {
+            let block = block?;
+            self.marked.remove(&truncated_hash(block.cid.hash()));
+        }
+
+        anyhow::Ok(())
+    }
+
+    // Remove marked keys from the database.
+    fn sweep(&mut self) -> anyhow::Result<()> {
+        let mut marked = HashSet::new();
+        mem::swap(&mut marked, &mut self.marked);
+        self.db.remove_keys(marked)
+    }
+
+    /// Starts the Garbage Collection loop.
+    ///
+    /// # Arguments
+    ///
+    /// * `depth` - Specifies how far back the full history should be maintained. Cannot be less
+    /// than chain finality.
+    /// * `interval` - GC Interval to avoid constantly consuming node's resources.
+    /// * `manual` - Whether or not the GC has to be triggered manually or automatically.
+    ///
+    /// NOTE: This currently does not take into account the fact that we might be starting the node
+    /// using CAR-backed storage with a snapshot, for implementation simplicity.
+    pub async fn gc_loop(&mut self, interval: Duration, manual: bool) -> anyhow::Result<()> {
+        loop {
+            match manual {
+                true => {
+                    let msg = self.gc_receiver.recv_async().await?;
+                    info!("running manual gc");
+                    let res = self.gc_workflow(manual).await;
+                    msg.send_async(res).await?;
+                    info!("finished manual gc run");
+                }
+                false => {
+                    self.gc_workflow(manual).await?;
+                    // Make sure we don't run the GC too often.
+                    time::sleep(interval).await;
+                }
+            }
+        }
+    }
+
+    async fn gc_workflow(&mut self, manual: bool) -> anyhow::Result<()> {
+        let depth = self.depth;
+        let tipset = self.chain_store.heaviest_tipset();
+        let current_epoch = tipset.epoch();
+        // Don't run the GC if there aren't enough state-roots yet. Manual GC overrides that.
+        if !manual || depth > current_epoch {
+            time::sleep(AVERAGE_BLOCK_TIME * (depth - current_epoch) as u32).await;
+            return anyhow::Ok(());
+        }
```

Comment:
I don't understand why manual GC overrides this check. If we aren't going to garbage collect anything, why run the garbage collector?

Reply:
The current manual GC CLI interface expects the thing to run immediately. I reckon we should either be able to run the GC immediately assuming the user knows what they are doing or remove it for good.

Reply:
As discussed in a different thread - we should either change the UX of the manual gc command or remove it for good. 
I didn't take forks into account, so we indeed need to employ the same rules for manual GC as we do for scheduled one.